src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -255,21 +302,97 @@ cdef class AioServer:
-        loop.create_task(_server_start(
-            self._server,
-            self._cq,
-            self._generic_handlers,
-        ))
+        cdef object server_started = self._loop.create_future()
+        self._serving_task = self._loop.create_task(self._server_main_loop(server_started))
+        # Needs to explicitly wait for the server to start up.
+        # Otherwise, the actual start time of the server is un-controllable.
+        await server_started
+
+    async def shutdown(self, grace):
+        """Gracefully shutdown the C-Core server.
+
+        Application should only call shutdown once.
+
+        Args:
+          grace: An optional float indicates the length of grace period in
+            seconds.
+        """
+        if self._status != AIO_SERVER_STATUS_RUNNING:
+            # The server either is shutting down, or not started.
+            return
+        cdef object shutdown_completed = self._loop.create_future()
+        cdef CallbackWrapper wrapper = CallbackWrapper(
+            shutdown_completed,
+            SERVER_SHUTDOWN_FAILURE_HANDLER)
+        # NOTE(lidiz) Without Py_INCREF, the wrapper object will be destructed
+        # when calling "await". This is an over-optimization by Cython.
+        cpython.Py_INCREF(wrapper)
+
+        # Starts the shutdown process.
+        # The shutdown callback won't be called unless there is no live RPC.
+        grpc_server_shutdown_and_notify(
+            self._server.c_server,
+            self._cq._cq,
+            wrapper.c_functor())
+        self._server.is_shutting_down = True
+        self._status = AIO_SERVER_STATUS_STOPPING
+
+        # Ensures the serving task (coroutine) exits.
+        try:
+            await self._serving_task
+        except _RequestCallError:
+            pass
+
+        if grace is None:
+            # Directly cancels all calls
+            grpc_server_cancel_all_calls(self._server.c_server)
+            await shutdown_completed
+        else:
+            try:
+                await asyncio.wait_for(asyncio.shield(shutdown_completed), grace)
+            except asyncio.TimeoutError:
+                # Cancels all ongoing calls by the end of grace period.
+                grpc_server_cancel_all_calls(self._server.c_server)
+                await shutdown_completed
+
+        # Keeps wrapper object alive until now.
+        cpython.Py_DECREF(wrapper)
+        grpc_server_destroy(self._server.c_server)
+        self._server.c_server = NULL
+        self._server.is_shutdown = True
+        self._status = AIO_SERVER_STATUS_STOPPED
+
+        # Shuts down the completion queue
+        await self._cq.shutdown()
 
-    # TODO(https://github.com/grpc/grpc/issues/20668)
-    # Implement Destruction Methods for AsyncIO Server
-    def stop(self, unused_grace):
-        pass
+    def __dealloc__(self):
+        if self._status != AIO_SERVER_STATUS_STOPPED:
+            _LOGGER.error('Server is not stopped while deallocation: %d', self._status)
```

Comment:
s/Server is not stopped while deallocation/`__dealloc__` called on running server/